src/typing_extensions.py

```diff
@@ -829,35 +853,73 @@ def __new__(cls, name, bases, ns, total=True):
                 required_keys.update(base.__dict__.get('__required_keys__', ()))
                 optional_keys.update(base.__dict__.get('__optional_keys__', ()))
+                readonly_keys.update(base.__dict__.get('__readonly_keys__', ()))
+                mutable_keys.update(base.__dict__.get('__mutable_keys__', ()))
+
+                if not getattr(base, "__other_keys__", True):
+                    if (
+                        set(own_annotations) -
+                        set(base.__dict__.get('__annotations__', {}))
+                    ):
+                        raise TypeError(
+                            "TypedDict cannot inherit from a TypedDict with "
+                            "other_keys=False and new fields"
+                        )
+                if (
+                    readonly
+                    and is_typeddict(base)
+                    and base is not _TypedDict
+                    and not getattr(base, "__readonly__", False)
+                ):
+                    raise TypeError(
+                        "read-only TypedDict cannot extend non-read-only "
+                        f"TypedDict {base}"
+                    )
 
             annotations.update(own_annotations)
             for annotation_key, annotation_type in own_annotations.items():
-                annotation_origin = get_origin(annotation_type)
-                if annotation_origin is Annotated:
-                    annotation_args = get_args(annotation_type)
-                    if annotation_args:
-                        annotation_type = annotation_args[0]
-                        annotation_origin = get_origin(annotation_type)
-
-                if annotation_origin is Required:
+                qualifiers = list(_get_typeddict_qualifiers(annotation_type))
+
+                if Required in qualifiers:
                     required_keys.add(annotation_key)
-                elif annotation_origin is NotRequired:
+                elif NotRequired in qualifiers:
                     optional_keys.add(annotation_key)
                 elif total:
                     required_keys.add(annotation_key)
                 else:
                     optional_keys.add(annotation_key)
+                if ReadOnly in qualifiers:
+                    if readonly:
+                        raise TypeError(
+                            "Using ReadOnly[] on a TypedDict with readonly=True "
+                            "is redundant")
+                    if annotation_key in mutable_keys:
+                        raise TypeError(
+                            f"Cannot override mutable key {annotation_key!r}"
+                            " with read-only key"
+                        )
+                    readonly_keys.add(annotation_key)
+                elif readonly:
+                    readonly_keys.add(annotation_key)
+                else:
+                    mutable_keys.add(annotation_key)
```

Comment:
Do you need to remove the key from readonly_keys too here? It might be inherited.

Reply:
This is actually already buggy for Required/NotRequired, even in CPython:

```
Python 3.13.0a2+ (heads/pep702:6ab4584ac4, Nov 26 2023, 17:26:43) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import *
>>> class A(TypedDict):
...     a: NotRequired[int]
... 
>>> class B(A):
...     a: Required[int]
... 
>>> B.__required_keys__
frozenset({'a'})
>>> B.__optional_keys__
frozenset({'a'})
```

I'll leave fixing that to a separate PR.

Reply:
(But I'll fix this for ReadOnly.)

Reply:
I think that example is illegal though. You can't change the required/notrequired on a key like that; it's banned by PEP-589. It's only once we have ReadOnly that this becomes an issue.

Reply:
Replied about this on CPython